src/type-extensions/proxy.ts

```diff
@@ -0,0 +1,164 @@
+import { BaseDecorationOptions, Decoration } from '../decoration'
+import { LuaReturn, LuaState, LuaType } from '../types'
+import MultiReturn from '../multireturn'
+import Thread from '../thread'
+import TypeExtension from '../type-extension'
+
+export interface ProxyDecorationOptions extends BaseDecorationOptions {
+    disableProxy?: boolean
+}
+
+export function decorateProxy(target: any, options?: ProxyDecorationOptions): Decoration<any, ProxyDecorationOptions> {
+    return new Decoration<any, ProxyDecorationOptions>(target, options || {})
+}
+
+class ProxyTypeExtension extends TypeExtension<any, ProxyDecorationOptions> {
+    private readonly gcPointer: number
+
+    public constructor(thread: Thread) {
+        super(thread, 'js_proxy')
+
+        this.gcPointer = thread.cmodule.module.addFunction((functionStateAddress: LuaState) => {
+            // Throws a lua error which does a jump if it does not match.
+            const userDataPointer = thread.cmodule.luaL_checkudata(functionStateAddress, 1, this.name)
+            const referencePointer = thread.cmodule.module.getValue(userDataPointer, '*')
+            thread.cmodule.unref(referencePointer)
+
+            return LuaReturn.Ok
+        }, 'ii')
+
+        if (thread.cmodule.luaL_newmetatable(thread.address, this.name)) {
+            const metatableIndex = thread.cmodule.lua_gettop(thread.address)
+
+            // Mark it as uneditable
+            thread.cmodule.lua_pushstring(thread.address, 'protected metatable')
+            thread.cmodule.lua_setfield(thread.address, metatableIndex, '__metatable')
+
+            // Add the gc function
+            thread.cmodule.lua_pushcclosure(thread.address, this.gcPointer, 0)
+            thread.cmodule.lua_setfield(thread.address, metatableIndex, '__gc')
+
+            thread.pushValue((self: any, key: unknown) => {
```

Comment:
And the keys that are also keywords? Do you think we shouldn't worry?

Reply:
You're going to have to elaborate there a bit. If you mean indexing a JS object by Lua keywords that's probably okay, you can do it in Lua using quotes anyway

Reply:
You are right, I forgot that because it's not possible when using the `:` syntax